src/plugin.ts

```diff
@@ -194,68 +83,96 @@ export default class DocgenPlugin {
         ...userCompilerOptions,
       };
     } else {
-      const { options } = getTSConfigFile(tsconfigPath);
-      compilerOptions = { ...compilerOptions, ...options };
+      const { options: tsOptions } = getTSConfigFile(tsconfigPath);
+      compilerOptions = { ...compilerOptions, ...tsOptions };
     }
 
-    const parser = docGen.withCompilerOptions(compilerOptions, docgenOptions);
-
-    compiler.hooks.make.tap(this.name, (compilation) => {
-      compilation.hooks.seal.tap(this.name, () => {
-        const modulesToProcess: Module[] = [];
-
-        compilation.modules.forEach((module: Module) => {
-          if (!module.built) {
-            debugExclude(`Ignoring un-built module: ${module.userRequest}`);
-            return;
-          }
-
-          if (module.external) {
-            debugExclude(`Ignoring external module: ${module.userRequest}`);
-            return;
-          }
-
-          if (!module.rawRequest) {
-            debugExclude(
-              `Ignoring module without "rawRequest": ${module.userRequest}`
-            );
-            return;
-          }
-
-          if (isExcluded(module.userRequest)) {
-            debugExclude(
-              `Module not matched in "exclude": ${module.userRequest}`
-            );
-            return;
-          }
+    this.options = options;
+    this.compilerOptions = compilerOptions;
+    this.docgenOptions = docgenOptions;
 
-          if (!isIncluded(module.userRequest)) {
-            debugExclude(
-              `Module not matched in "include": ${module.userRequest}`
-            );
-            return;
-          }
+    // Maintain the doc gen parser here to save time later
+    this.parser = docGen.withCompilerOptions(compilerOptions, docgenOptions);
+  }
 
-          debugInclude(module.userRequest);
-          modulesToProcess.push(module);
-        });
+  apply(compiler: webpack.Compiler): void {
+    const pluginName = "DocGenPlugin";
 
-        const tsProgram = ts.createProgram(
-          modulesToProcess.map((v) => v.userRequest),
-          compilerOptions
-        );
+    const { exclude = [], include = ["**/**.tsx"] } = this.options;
+    const isExcluded = matchGlob(exclude);
+    const isIncluded = matchGlob(include);
 
-        modulesToProcess.forEach((m) =>
-          processModule(parser, m, tsProgram, {
-            docgenCollectionName,
-            setDisplayName,
-            typePropName,
+    compiler.hooks.compilation.tap(
+      pluginName,
+      (compilation, { normalModuleFactory }) => {
+        compilation.dependencyTemplates.set(
+          // eslint-disable-next-line
+          // @ts-ignore TODO: Figure out why this isn't allowed
+          DocGenDependency,
+          new DocGenDependency.Template({
+            parser: this.parser,
+            docgenOptions: this.docgenOptions,
```

Comment:
This is a bit risky and won't work when multiple instances of this plugin with difference options are used.